engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/TypeOfNode.java

```diff
@@ -193,4 +103,150 @@ Object doAny(Object value) {
+    abstract Object execute(Object value);
+
+    @Specialization(guards = {"findInterop(value, interop).isArray()"})
+    Type doPolyglotArray(Object value, @CachedLibrary(limit = "3") InteropLibrary interop) {
+      return EnsoContext.get(this).getBuiltins().array();
+    }
+
+    @Specialization(guards = {"findInterop(value, interop).isString()"})
+    Type doPolyglotString(Object value, @CachedLibrary(limit = "3") InteropLibrary interop) {
+      return EnsoContext.get(this).getBuiltins().text();
+    }
+
+    @Specialization(guards = {"findInterop(value, interop).isNumber()"})
+    Type doPolyglotNumber(Object value, @CachedLibrary(limit = "3") InteropLibrary interop) {
+      Builtins builtins = EnsoContext.get(this).getBuiltins();
+      if (interop.fitsInInt(value)) {
+        return builtins.number().getInteger();
+      } else if (interop.fitsInDouble(value)) {
+        return builtins.number().getDecimal();
+      } else {
+        return EnsoContext.get(this).getBuiltins().number().getNumber();
+      }
+    }
+
+    @Specialization(guards = {"findInterop(value, interop).isDateTime()"})
+    Type doDateTime(Object value, @CachedLibrary(limit = "3") InteropLibrary interop) {
+      return EnsoContext.get(this).getBuiltins().dateTime();
+    }
+
+    @Specialization(guards = {"findInterop(value, interop).isTimeZone()"})
+    Type doTimeZone(Object value, @CachedLibrary(limit = "3") InteropLibrary interop) {
+      EnsoContext ctx = EnsoContext.get(this);
+      return ctx.getBuiltins().timeZone();
+    }
+
+    @Specialization(guards = {"findInterop(value, interop).isDate()"})
+    Type doDate(Object value, @CachedLibrary(limit = "3") InteropLibrary interop) {
+      EnsoContext ctx = EnsoContext.get(this);
+      return ctx.getBuiltins().date();
+    }
+
+    @Specialization(guards = {"findInterop(value, interop).isTime()"})
+    Type doTime(Object value, @CachedLibrary(limit = "3") InteropLibrary interop) {
+      EnsoContext ctx = EnsoContext.get(this);
+      return ctx.getBuiltins().timeOfDay();
+    }
+
+    @Specialization(guards = "findInterop(value, interop).isDuration()")
+    Type doDuration(Object value, @CachedLibrary(limit = "3") InteropLibrary interop) {
+      EnsoContext ctx = EnsoContext.get(this);
+      return ctx.getBuiltins().duration();
+    }
+
+    @Specialization(guards = {"findInterop(value, interop).isNone()"})
+    Object doMetaObject(Object value, @CachedLibrary(limit = "3") InteropLibrary interop) {
+      try {
+        return interop.getMetaObject(value);
+      } catch (UnsupportedMessageException e) {
+        CompilerDirectives.transferToInterpreter();
+        Builtins builtins = EnsoContext.get(this).getBuiltins();
+        throw new PanicException(builtins.error().makeCompileError("invalid meta object"), this);
+      }
+    }
+
+    static InteropType findInterop(Object value, InteropLibrary interop) {
+      return InteropType.find(value, interop);
+    }
+
+    enum InteropType {
```

Comment:
`findInterop` seems like a good simplification. But can we move `InteropTYpe` to some Utils class such that it can also be used by specializations in different classes?

Reply:
Right now the `InteropType` is closely related to the specializations in the `TypeOfNode` - where else would you use it? I cannot imagine it would fit.